runtime/src/main/java/org/corfudb/runtime/collections/StreamManager.java

```diff
@@ -0,0 +1,225 @@
+            throw new StreamSubscriptionException(
+                    "subscribe: Max " + MAX_SUBSCRIBERS + " hit! Consider consolidating "
+                            + streamListener);
+        }
+        if (subscriptions.containsKey(streamListener)) {
+            throw new StreamSubscriptionException(
+                    "StreamManager:subscriber already registered "
+                            + streamListener + ". Maybe call unregister first?");
+        }
+        log.info("StreamManager::subscribe {}, startAddress {}, namespace {}, tables {}",
+                streamListener, startAddress, namespace, tablesOfInterest.toString());
+        SubscriptionThread st = new SubscriptionThread(this, streamListener,
+                namespace, tablesOfInterest, startAddress);
+        subscriptions.put(streamListener, st);
+        st.start();
+    }
+
+    /**
+     * Unsubscribe a prior subscription.
+     *
+     * @param streamListener Client listener.
+     */
+    public synchronized void unsubscribe(@Nonnull StreamListener streamListener) {
+        SubscriptionThread st = subscriptions.remove(streamListener);
+        if (st == null) {
+            log.warn("StreamManager::unsubscribe has no context for {}", streamListener.toString());
+            return;
+        }
+        st.shutdown.set(true);
+        st.interrupt();
+        log.info("Unsubscribed StreamListener {}", streamListener.toString());
+    }
+
+    class SubscriptionThread extends Thread {
+        /**
+         * The callback context
+         */
+        private final StreamListener listener;
+
+        /**
+         * Parent class to manage errors.
+         */
+        @Getter
+        private final StreamManager streamManager;
+
+        /**
+         * Namespace of interest.
+         */
+        @Getter
+        private final String namespace;
+
+        /**
+         * Tables of interest.
+         *
+         * Map of (Stream Id - TableSchema)
+         */
+        @Getter
+        private final Map<UUID,
+                TableSchema<? extends Message, ? extends Message, ? extends Message>> tablesOfInterest;
+
+        /**
+         * Starting address.
+         */
+        private final long startAddress;
+
+        public AtomicBoolean shutdown;
+
+        /**
+         * A reference to the underlying stream
+         */
+        @Getter
+        private final IStreamView txnStream;
+
+        public void run() {
+            try {
+                log.info("{}:Seeking txStream to {}", listener.toString(), startAddress + 1);
+                txnStream.seek(startAddress + 1);
+            } catch (Throwable throwable) {
+                log.warn("{}:onError at seek to {}", listener.toString(), startAddress + 1);
+                listener.onError(throwable);
+                streamManager.unsubscribe(listener);
+                return;
+            }
+
+            try {
+                while (!shutdown.get()) {
```

Comment:
If I understand it properly what you are trying to do is to have scheduler which will run your task periodically.
You can use java scheduler and it simplifies code. You will not need to check `shutdown` variable, no need for thread.sleep, catching exceptions to stop the thread, and so on.

Please take a look https://github.com/CorfuDB/CorfuDB/compare/simplifyStreaming...simplifyStreaming-scheduler

Reply:
Also you can have as many threads as you need (scheduler threadfctory)

Reply:
Your implementation seems to be creating one thread factory in every thread. Why?

Reply:
Please see https://stackoverflow.com/a/50077162/4513389 for why this is an overkill for this use case.

Reply:
It's just a dirty example, you can put scheduler and factory into places where the should be

Reply:
> Please see https://stackoverflow.com/a/50077162/4513389

A very weird explanation, not sure what that guy tried to say :)

Reply:
scheduler type model with a thread pool is exactly what we have currently (please check out the code that was removed by this PR) - the problem is when you have varying rates of consumption we have multiple tasks executing in the same thread and the heap (from thread local data) grows too quickly for JVMs with low heap space to handle. This is the very bug which prompted this simplification in the first place.

Adding back a scheduler and factory model where multiple tasks are going to execute on the same thread will simply bring back the very problem we are trying to fix.

Reply:
I don't believe that scheduler could be a problem in this case. I don't see how it could be possible in java when "the heap (from thread local data) grows too quickly" it sounds a bit odd, also if you just replace scheduler by threads I'm pretty sure it doesn't improve anything just by definition it can't improve performance/resource utilization/consumption.
So, it's up to you, but replacing scheduler by threads not a good idea imo (usually threads gett replaced by scheduler)

Reply:
It's not a problem with scheduler, but a problem with having multiple tasks executing on the same thread pulling different set of log data objects from stream layer.
So to avoid that we really need only have one task running on one thread. Given this requirement a scheduler is really pointless which is what the stackoverflow link above is trying to explain. Once we move to stream tagged selective streaming we will now be able to selectively pull items, at which point it might make sense to consider multiple tasks on a thread or thread factory or pools or even the reactive-streams framework if it makes sense. Technologies have to be applied to serve use cases not the other way around right?

Reply:
Let's clarify.
Invariants:
 - a thread can execute one task at a time, it's not possible to execute many tasks by one thread at a time
 - your task must not keep any state in the thread, you shouldn't use ThreadLocal variables if you want to use a common pool of threads.

If I understand it properly, we have some ThreadLocal variables which bring side effects?
If that true, then it's poor design/implementation in Corfu and we are trying to hack it here and make it work.

If I'm wrong then ok, but I don't see how using a scheduler can be in any way worse than using threads. I would prefer to fix a problem in the design than replace a better solution by a more complex and fragile solution

Reply:
Here is a simpler question - why do you need a thread scheduler/factory if a thread only needs to ever execute one task?